### PR TITLE
NEXUS-4640: The Parent-O-Matic device

### PR DESCRIPTION
**Pull request by cstamas**
As Peter's testing against generated snapshot repository with 13000 artifacts revealed an interesting results (snapshot removal using 1.9.2.3 takes 3(!) minutes, but the "post processing" is 20 minutes, hence to complete the task you need to wait cca 23 minutes), the need for a special device emerged.

And here it is: Parent-O-Matic!

The problem with Snapshot remover was actually a lame "optimization" added in there, that seemed correct in "laboratory" circumstances (and ITs), but resulted in "anti-optimization" in real life: there was a Map that gathered which paths need to have "rebuild maven metadata" regeneration applied. On large repositories, this path ended up with all the existing paths in repository, and even OOMed Nexus. 

Finally, this change does not makes previous (already merged) pull request also related to NEXUS-4640 obsolete. While it also fixed major issues, it's fix covered so to tell just 10% of the problem, while this fix covers the rest of it, and even OOM related issue (the one linked to NEXUS-4640).

### CODE REVIEW

File: nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/walker/ParentOMatic.java [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+    {
+        // rule a: if parent is marked already, do not mark the child
+        if ( isParentMarked( changedNode ) )
+        {
+            changedNode.getPayload().setMarked( false );
+            return changedNode.getParent();
+        }
+
+        // rule b: if this parent's all children are marked, mark parent, unmark children
+        if ( isParentAllChildMarkedForRuleB( changedNode ) )
```
Comment (nabcos) [inline nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/walker/ParentOMatic.java:264]: this should be called 'siblings' IIUC?

**Comment by cstamas** [comment]
CI build scheduled:
https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/103/

**Comment by adreghiciu** [comment]
+1

**Comment by cstamas** [comment]
CI voted :cake:
https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/103/

**Comment by peterlynch** [comment]
+1 thx for ci build

File: nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/walker/ParentOMatic.java [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+            final List<Node<Payload>> children = parent.getChildren();
+
+            if ( children.size() < 2 )
+            {
+                return false;
+            }
+
+            for ( Node<Payload> child : children )
+            {
+                if ( !child.getPayload().isMarked() )
```
Comment (nabcos) [inline nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/walker/ParentOMatic.java:358]: not sure I understand correctly:
- if `node` is the only child, it is treated as not marked and returns false
- if there are more children, it returns false if one of them is not marked

Doesn't that always return false, assuming that `node` has a non-marked payload as assumed in the first condition?

Comment (cstamas) [inline nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/walker/ParentOMatic.java:358]: The `node` parameter is always marked, just see how this method is invoked (entry point is addAndMarkPath()).

This code returns `true` in following case:

``` java
addAndMarkPath("/parent/sub1");
addAndMarkPath("/parent/sub2");
```

Second invocation will return `true` here.

File: nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/walker/ParentOMatic.java [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+    }
+
+    /**
+     * Returns true if parent exists (passed in node is not ROOT), and parent's all children are marked (their
+     * {@link Payload#isMarked()} is {@code true} for all of them.
+     * 
+     * @param node
+     * @return
+     */
+    protected boolean isParentAllChildMarkedForRuleB( final Node<Payload> node )
```
Comment (bdemers) [inline nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/walker/ParentOMatic.java:343]: confusing method name,  There is no matching "RuleA" method

File: nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/walker/ParentOMatic.java [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+ * <p>
+ * This class also "optimizes" the tree size to lessen memory use. This "optimization" can be turned off, see
+ * constructors.
+ * <p>
+ * This class makes use of {@link Node} to implement the tree hierarchy.
+ * 
+ * @author cstamas
+ * @since 1.10.0
+ */
+public class ParentOMatic
```
Comment (bdemers) [inline nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/walker/ParentOMatic.java:59]: As much as i like anything named *-O-Matic,  I think this needs a more descriptive name.

File: nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/walker/ParentOMatic.java (hunk 2 above)
Comment (nabcos) [inline nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/walker/ParentOMatic.java:358]: But `if ( children.size() < 2 ) { return false }` that means you treat `node` as not marked in this condition?

**Comment by nabcos** [comment]
Is there really no usable treenode class in the JRE other than the Swing one? otherwise okay aside of naming (although I think we should let all nexus classes end with -OMatic, just for style)

+1

**Comment by cstamas** [comment]
Thanks Ben! It's going in.

File: nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/walker/ParentOMatic.java (hunk 2 above)
Comment (nabcos) [inline nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/walker/ParentOMatic.java:358]: I still think this is a bug:

```
cn.addPath( "/parent" );
cn.addAndMarkPath( "/parent/child" );
```

will return `false` for `isParentAllChildMarked...` here. Is that intended?

**Comment by nabcos** [comment]
I still don't think ParentOMatic is a useful description of what this is. I will never find it again half a year from now, so we should really change that.

**Pull request merged**